core/include/io.h

```diff
@@ -273,4 +273,37 @@ static inline void io_clrsetbits8(vaddr_t addr, uint8_t clear_mask,
 	io_write8(addr, (io_read8(addr) & ~clear_mask) | set_mask);
 }
 
+/*
+ * Poll on a IO memory content or timeout
+ *
+ * @addr is the address of the memory cell accessed
+ * @val represents the val  of the memory cell accessed
+ * @cond represents the condition to get the correct value
+ * @delay_us represents the read interval in mircorseconds
+ * @_timeout_us represents the time-out period in microseconds
+ *
+ * @return true means time-out, false means got right value
+ */
+#define IO_READ32_POLL_TIMEOUT(addr, val, cond, delay_us, _timeout_us) \
+	({ \
+		uint64_t _timeout = 0; \
+		uint64_t _delay_us = delay_us; \
+		bool _flag = true; \
```

Comment:
I'd prefer `__ret` or `__rc`. We should probably prefix the locals with `__` instead.